.circleci/config.yml

```
@@ -221,14 +206,36 @@ references:
       openssl aes-256-cbc -md md5 -d -in desktop/resource/calypso/secrets.json.enc -out config/secrets.json -k "${CALYPSO_SECRETS_ENCRYPTION_KEY}"
       openssl aes-256-cbc -md md5 -d -in desktop/resource/certificates/mac.p12.enc -out desktop/resource/certificates/mac.p12 -k "${CALYPSO_SECRETS_ENCRYPTION_KEY}"
       openssl aes-256-cbc -md md5 -d -in desktop/resource/certificates/win.p12.enc -out desktop/resource/certificates/win.p12 -k "${CALYPSO_SECRETS_ENCRYPTION_KEY}"
-  desktop-restore-source-cache: &desktop-restore-source-cache
-    name: Restore desktop cache
+  desktop-restore-yarn-cache: &desktop-restore-yarn-cache
+    name: Restore yarn cache
+    keys:
+      - v6-desktop-yarn-{{ .Environment.CIRCLE_JOB }}-{{ arch }}-{{ checksum ".nvmrc" }}-{{ checksum "yarn.lock" }}
```

If you want to be more specific, we should add `.yarnrc` as well to the hash. It contains yarn version among other things that can affect the cache.

---

Actually I'm wrong. This cache is just a set of downloaded packages, as far as I know is not related to `yarn.lock` or `.nvmrc`. Maybe use only `.yarnrc` in case a different yarn version has a different, incompatible cache format?

---

ah, I see. I also just noticed there's existing (unused) logic for Calypso that keys the cache of `.nvmrc` and `yarn.lock`. Perhaps we can just go with those keys here as well?  

If `.nvmrc` changed wouldn't `yarn.lock` change as well (i.e. no need to specify both)?

---

`.nvmrc`, `yarn.lock` and `.yarnrc` can change independently.

But it is irrelevant for _this_ cache. This cache is just a copy of the packages downloaded from internet, unrelated to the version of node or yarn used to download them. I'd just restore it always (ie, not use `.nvmrc`, `yarn.lock` and `.yarnrc` for the keys). 

The worst case scenario is that some packages are not cached and has to be downloaded again. But if a package is already in the cache, it will work even if it was requested with a different `.nvmrc`, `yarn.lock` or `.yarnrc`.

---

Unfortunately as far as I can tell the packages-build step isn't selective - it will rebuild all packages regardless of whether any of them might have been built prior (unlike yarn, which is smart about pulling from the yarn cache if a yarn cache is specified). The best we can do is a blanket verification of whether _anything_ in the packages directory has changed, and rebuild the full set if so.

If the source code in the packages directory is unchanged, but the yarn/npm/node version changed since the cache was last saved, it seems we'd want to trigger a full cache bust/rebuild, yeah?

---

100% agree, but this comment was not about the packages build step... 

The cache I'm talking about is about the cache `v6-desktop-yarn-...` which contains `.cache/yarn`, which is just a mirror of the packages tgzs downloaded from a npm registry. It is safe to cache "for ever", it is not dependant on node version or yarn version.

The key of this cache can be:

```suggestion
      - v6-desktop-yarn-{{ .Environment.CIRCLE_JOB }}-{{ arch }}
```

---

Ah, I see - didn't know that, makes sense. In that case the proposed change sounds reasonable. However I think we'll still want to at least _occasionally_ bust the cache to refresh its contents (to keep it reasonably up-to-date with the npm registry). At the very least, I think both saving and restoring the cache should include `.nvmrc` for the occasional refresh (even though `.nvmrc` is not a dependency of the cache). I would suggest leaving `.nvmrc` in at least, which ensures that the cache is busted periodically.

Updated in 42861220baf4bb34131438dfea96655be3abac46.